### ISSUE TITLE
Enhance the 'Request a Change to this Listing' feature

### ISSUE THREAD

**Issue by kiyote33** [code4recovery/12-step-meeting-list#362]
---
name: Enhance Meeting Detail Feedback System

about: Replace or augment the existing Feedback feature of the Meeting Detail screen (single-meetings.php) with an enhanced New and Change Request input screen. Similar to the old feedback screen, the new/change request screens would remain hidden until activated by a user, only then replacing the map view in the right side column of the Detail page. The input fields would be transformed by the user in a  controlled fashion prior to submitting the form data, ensuring that the website administrator receives a full set of accurate and auditable data with which to update the Meeting Information. The submitted 
data would be parsed and injected into an html table within an email, with changes highlighted in Red text for easy analysis and action by the system administrator. 

title: 'Enhance Change Request System'
labels: 'new feature/enhancement'
assignees: ''

---

**Is your feature request related to a problem? Please describe.**
Yes. The current Feedback feature relies on the average user to have more knowledge of the Meeting attributes than is reasonable to expect. Mistakes and misunderstandings are common-place which could easily be avoided by providing a formatted solution to guide the user input, giving a consistent, auditable, and accurate view of what the Requestor is wanting added or changed to the 12 Step Meeting List solution. 

**Describe the solution you'd like**
Ideally, all meeting attributes are available in the form of input controls, using radio buttons, checkboxes, and dropdown lists whenever possible. Free form text controls should suffice for the rest, with all controls being managed with validation mechanisms when possible. Hook into the existing Feedback feature activation mechanism, possibly giving system administrators the choice of overriding the simple Feedback form with the more enhanced Change Request form.

**Describe alternatives you've considered**
An alternative could be to incorporate this as a new feature instead of an extension to an existing feature. Possibly create a separate stand-alone plug-in using Popups to configure the Request prior to submitting.

**Additional context**
See screenshots MeetingDetailScreen1.png, MeetingDetailScreen2.png, MeetingRequestSubmitted.png, and ChangeRequestEmail.png.
![MeetingDetailScreen1](https://user-images.githubusercontent.com/57125427/120769430-59984c80-c4da-11eb-9f28-fba7ff9ce189.png)

 shortly).
Focus above is on the Change Meeting Request, but the feature must include the associated Add New Meeting Request and Remove Meeting screens as well.

**Comment by kiyote33** [comment]
![MeetingDetailScreen2](https://user-images.githubusercontent.com/57125427/120821880-11495080-c513-11eb-9c92-9a9a35a8f07f.png)
![MeetingRequestSubmitted](https://user-images.githubusercontent.com/57125427/120821895-15756e00-c513-11eb-9188-32681690dda7.png)
![ChangeRequestEmail](https://user-images.githubusercontent.com/57125427/120821916-19a18b80-c513-11eb-8a96-26d2f0aa1c94.png)

**Comment by kiyote33** [comment]
Hmmm. Didn't mean to close....

**Comment by tech2serve** [comment]
Hey @kiyote33 

Thank you for this great write up, and taking a stab! Please shout if you have any questions or we can be of help.

Tim

**Comment by kiyote33** [comment]
Thanks Tim. I’m sort of new to the direct GitHub experience, having over the years depended on the graphical interfaces built into Visual Studio to manage source code. I’ll need to spend some time studying it to get up to speed with everyone else. Hopefully I’ll get there without causing too much pain and suffering to others in the community. Please feel free to pass on any suggestions or corrections that might help with that process.

 

This is also my first time working with the code in a WordPress site plug-in. So I’m learning about the pros & cons of different implementations.  The prototype I’ve developed for this feature has major modifications to both the single-meetings.php and ajax.php files that I’ve submitted for review. On my dev site I drop the Template file into my themes child folder and that works well. But putting an override hook for the tsml_ajax_feedback function into the theme child folder functions.php file didn’t work for me, so for expediency I ended up editing that routine in the Includes ajax.php file directly. 

 

In the prototype I’ve attempted to maintain the same feedback functionality currently available, while adding the change/add/delete request features. I’m thinking that may make it a good candidate to simply replace the existing feedback code so we are not needing to maintain two sets of code doing the same thing.  I’m certainly open to suggestions and what others think is the best way to handle this. 

 

Regards…

Wayne

**Comment by kiyote33** [comment]
I've moved the code work on this item from the prototype stage through code complete and unit testing phases.  Code is checked in with two files (single-meetings.php and functions.php) in the Child Theme folder for anyone wanting to test the feature.  My demo site at https://parkercanada.info is also available for anyone just wanting a quick look at the current state of this feature enhancement.  Feedback is welcome.

**Comment by tech2serve** [comment]
Hey @kiyote33--

Thank you for the demo site! I took a quick look, and will ask others in the group to do so as well. I think a lot of users would like something like this.

 I've not had the chance yet to look at the PR, although we typically try to treat PRs as actionable and ready to merge if the reviewer agrees with it and hasn't asked for any changes. 

I'll strive to look at this in-depth this week. We are focused on getting a beta version out with better support for online meetings. That is our first priority.

**Comment by kiyote33** [comment]
Hi Tim,
I added the code for admin_import.php so users can choose between the Legacy and Enhanced feedback screens. Just a few tweaks there and adding a new variable in variables.php.  I'll freeze the code now till I hear from you.  

Regards...

**Comment by kiyote33** [comment]
This feature released as a plugin "12-step-meeting-list-feedback-enhancement v3.12". Has dependency on the 12 Step Meeting List plugin, version 3.12.0.
